primitives/src/predicate.rs

```diff
@@ -0,0 +1,333 @@
+use crate::{Claim, ClaimType, Rule, RuleType};
+use codec::{Decode, Encode};
+
+use sp_std::prelude::*;
+
+/// Context using during an `Predicate` evaluation.
+///
+/// # TODO
+///  - Use a lazy access to claims. It could be part of the optimization
+///  process of CDD claims.
+#[derive(Encode, Decode, Clone, Debug, Default)]
+pub struct Context {
```

Comment:
"Context" is not really a lucky choice of word here. Predicates map values of their variables to true or false. Quantifiers map predicates to true or false. Those two work at a different level. Usually by context people mean something different than either values or predicates. It's actually more like assignment of values to free variables and a list of axioms. I'm not going to nag on this but this is the thing.

In addition, "and", "or" and "not" map propositions (things that are either true or false) to propositions, rather than values to true/false. But that's OK in your context because by `and` you mean "apply these two predicates to the same subject". A bit confusing though.

I suggest renaming `Context` as `Value` or `Subject`. Also references to quantifiers like `any` or `exists` better be removed because there aren't any quantifiers really, just predicates again.

Reply:
The idea behind "Context" is that "rule" defines the filters to apply on fetching claims. That is why I discarded other names like 'Value'. 
Right now it stores values in memory, but there is an optimization task on Jira, in which we could explore the possibility of use directly the iterator extracted from fetching phase.
As a context, we could also add other information that future predicates would need.

Reply:
In that case it's better not to call functions taking context to true/false predicates. Would calling them `Rule`s be possible? That would be better for things like joining predicates ("and") and alternating predicated ("or") as that would remove the unwanted connotations with predicate logic which this rule system is not.